database-mongodb/MySQL/testDatabaseQueryEfficiency.js

```diff
@@ -0,0 +1,567 @@
+  return response[0].itemImages;
+};
+
+const extractMultipleItems = function(response) {
+  const result = [];
+  for (let i = 0; i < response.length; i++) {
+    result.push(response[i].itemImages);
+  }
+  return result;
+};
+
+const makeActualCall = function(i, dataArray, queryTimeArray, queryHandler, dataSource, database, extractFunction, secondDataSource = []) {
+  queryTimeArray[i] = [new Date().getTime()];
+  queryHandler(dataSource[i], database, secondDataSource[i])
+    .then((response) => {
+      dataArray.push(extractFunction(response));
+      queryTimeArray[i][1] = new Date().getTime();
+      if (i < numberOfRandomQueries - 1) {
+        setTimeout(
+          makeActualCall.bind(
+            null,
+            i + 1,
+            dataArray,
+            queryTimeArray,
+            queryHandler,
+            dataSource,
+            database,
+            extractFunction,
+            secondDataSource
+          ),
+          10
+        );
+      } else {
+        handleTests();
+      }
+    })
+    .catch((err) => {
+      console.log(err);
+    });
+};
+
+const makeActualUpdateCall = function(i, dataArray, queryTimeArray, queryHandler, dataSource, database, extractFunction, queryHandlerTwo) {
+  queryHandler(dataSource[i], database)
+    .then((response) => {
+      const { itemImages } = response[0];
+      if (itemImages) {
+        const splitItemImages = itemImages.split('XXX');
+        const newItemImages = splitItemImages.join('OOO');
+        queryTimeArray[i] = [new Date().getTime()];
+        return queryHandlerTwo(dataSource[i], database, newItemImages);
+      } else {
+        throw new Error('itemId not found');
+      }
+    })
+    .then((response) => {
+      dataArray.push(extractFunction(response));
+      queryTimeArray[i][1] = new Date().getTime();
+      if (i < numberOfRandomQueries - 1) {
+        setTimeout(
+          makeActualUpdateCall.bind(
+            null,
+            i + 1,
+            dataArray,
+            queryTimeArray,
+            queryHandler,
+            dataSource,
+            database,
+            extractFunction,
+            queryHandlerTwo
+          ),
+          10
+        );
+      } else {
+        handleTests();
+      }
+    })
+    .catch((err) => {
+      console.log(err);
+    });
+};
+
+const makeActualCallSync = function(i, dataArray, queryTimeArray, queryHandler, data, database, extractFunction) {
+  queryTimeArray[i] = [new Date().getTime()];
+  queryHandler(data, database)
+    .then((response) => {
+      dataArray.push(extractFunction(response));
+      queryTimeArray[i][1] = new Date().getTime();
+      if (i === numberOfRandomQueries - 1) {
+        handleTests();
+      }
+    })
+    .catch((err) => {
+      console.log(err);
+    });
+};
+
+const createResponseData = [];
+const createResponseQueryTimes = [];
+const testMySQLCreate = function(queryHandler = insertRecord) {
+  console.log('create started');
+
+  setTimeout(
```

Comment:
what is the advantage of using setTimeout to call `makeActualCall` the first time since there is already a set timeout within `makeActualCall`?

Reply:
No advantage. I just have a degree in Chemistry and many years experience in scientific research. The idea is to try to avoid random or systematic sources of error whenever possible. Since all subsequent calls use setTimeout, I feel the first call should, too.